### PR TITLE
Add documentation for rebasing.

### PR DESCRIPTION
**Pull request by jeromekelleher**
This ended up being quite long, but hopefully takes some of the mystery out of rebasing and squashing.

What do you think @petrelharp, @andrewkern?

### CODE REVIEW

File: docs/development.rst [hunk 1]
```diff
@@ -50,19 +50,277 @@ GitHub workflow
...
+    error: failed to push some refs to 'git@github.com:mcveanlab/stdpopsim.git'
+    hint: Updates were rejected because the tip of your current branch is behind
+    hint: its remote counterpart. Integrate the remote changes (e.g.
+    hint: 'git pull ...') before pushing again.
+    hint: See the 'Note about fast-forwards' in 'git push --help' for details.
+
+**DO NOT LISTEN TO GIT IN THIS CASE!** Git is giving you is **terrible advice**
+which will mess up your branch. Instead, force-push::
+
+    $ git push -f origin better-rebase-docs
```
Comment (arundurvasula) [inline docs/development.rst]: Here I might make it clear that you should push to the branch that you are working from (i.e. not master). Might be able to add a reference to "topic_branch_name" as referenced above in  `git checkout -b topic_branch_name`

Comment (petrelharp) [reply docs/development.rst]: and also make extra clear you are force-pushing to *your* fork.

File: docs/development.rst [hunk 2]
```diff
@@ -50,19 +50,277 @@ GitHub workflow
...
+just make a diff of your current changes, and apply these in a single commit::
 
-        $ git push -f origin topic_branch_name
+    $ git diff upstream/master > changes.patch
+    $ git reset --hard upstream/master
+    $ patch -p1 < changes.patch
+    $ git commit -a
+    $ git push -f origin my_topic_branch
 
+This will completely remove **all** of your old commits, so beware!
```
Comment (petrelharp) [inline docs/development.rst]: perhaps we should suggest doing this on a separate branch first, to check that it works ok? Like

```
git checkout -b test_branch upstream/master
patch -p1 < changes.patch
git commit -a
# check things work
git checkout my_topic_branch
git reset --hard test_branch
git push -f origin my_topic_branch 
```

Comment (jeromekelleher) [reply docs/development.rst]: Yeah, good idea. This is very much the nuclear option, needs to be treated with care.

File: docs/development.rst (hunk 1 above)
Comment (jeromekelleher) [reply docs/development.rst]: Good feedback thanks, will do.

**Comment by jeromekelleher** [comment]
Thanks for the feedback, very helpful. I'll make another pass through and ping you all.

**Comment by jeromekelleher** [comment]
OK, I've made another pass through this and hopefully addressed comments.

**Comment by jeromekelleher** [comment]
@andrewkern, can you merge this if you're happy please?

**Pull request merged**